### PR TITLE
Database backend

### LINKED ISSUE #18: Failed to extract post: Url domain not supported
Hi,
temp

I was trying to use your GUI based downloader, and i wanted to download the top posts of all time from the subreddit www.reddit.com/r/learnprogramming 

I added the subreddit into the subreddit list. Checked download subreddit. I have attached a screenshot of my settings as well. 

After all that, i get the error : Failed to extract post: Url domain not supported. 
temp 2

Could you please help me out, let me know what im doing wrong. They are almost all text posts with some posts containing links. No images, no videos etc. Thank you so much for your help and for creating this program.

### LINKED ISSUE #22: Scheduling a download
More of an enhancement idea than an issue.  

Is there a way or plans to schedule a download at regularly set intervals?  I find that some users are deleting posts after they have been up for less than a day and I would like to capture it if I am not at the computer.

### LINKED ISSUE #30: Allow us to disable video sites
A window to allow us to enable which sites we want videos to be downloaded from and which we want disabled since some sites do not work or if we do not want certain sites to download.
The new video downloader does not work 100%  as if I try download a user and if for example they have p0rnhub link it does not download thus going into an infinite page download.But what if I want to disable that site all together? Removing said site from supported_video_sites .txt does not work as it just say cannot find extractor but it will still try to download it anyway.
something like:

```
-----------------------------------|III|-----------------------------
     sites enabled                 |III|          availables sites           
-----------------------------------|III| ----------------------------
Youtube                            |III|  P0rnhub
Vimeo                              |III|  AMC
etc                                |III|  etc


```

### LINKED ISSUE #46: Unable to save user list
I am having a lot of trouble saving the user list.
If I add users then immediately save, it's all good.
If I add users and then start the download, then one of 2 things happens:
1) the final download to complete is a success (an image on a host supported by the app) -> then all is good, the user list and progress can be saved.

2) if the final download to process is a failure (unsupported type) -> unable to save, any users added if not saved before starting the download are lost. The modification date on the save file is not changed.

When  scenario 2 happens, the "download" button remains stuck, clicking it to stop does nothing. Attempting to save via the file menu gives a failure message.

The biggest issue with this is not only are users added to the list discarded if not saved before starting the download, but the same content is downloaded over and over again until a scenario # 1 happens and a save can be done properly.

the log shows this when attempting to save, also URL to screen capture of pop up error.

`{"levelname": "ERROR", "name": "DownloaderForReddit.DownloaderForReddit.Persistence.ObjectStateHandler", "filename": "ObjectStateHandler.py", "module": "ObjectStateHandler", "funcName": "save_pickled_state", "lineno": 129, "message": "Unable to save to save_file", "asctime": "12/07/2018 04:32:02 PM", "exc_info": "Traceback (most recent call last):\n  File \"DownloaderForReddit\\Persistence\\ObjectStateHandler.py\", line 121, in save_pickled_state\n  File \"shelve.py\", line 124, in __setitem__\nTypeError: can't pickle Content objects", "save_file_location": "C:\\Users\\Administrator\\AppData\\Roaming\\SomeGuySoftware\\DownloaderForReddit\\save_file"}`

[url](https://i.imgur.com/jVF4wp1.png)

### LINKED ISSUE #47: Migrating to sqlite support?
I know this is on your todo list. I've put some thought into this too and I'm wondering how I can help. I don't really have time right now, but maybe if we can split it up into small chunks it'll be easier to do.

## To ORM or not to ORM?

Should we use the stdlib sqlite3 package or an ORM?

* https://docs.python.org/3/library/sqlite3.html
* Peewee ORM http://docs.peewee-orm.com/en/latest/
* PonyORM https://ponyorm.com/
* SQLAlchemy https://www.sqlalchemy.org/

## Initial migration path

1. Hard migrations – we have a one time migration. One and done. This would mean everything has to get converted to sqlite
2. Gradual migration – migrate one model at a time. RedditObjects, Redditors, Redditor preferences, Subreddits, Subreddit preferences, system preferences, etc.

## Maintenance

Deleting old data? Vacuums? Schema migrations?

### LINKED ISSUE #72: Allow downloads from commented imgur/erome links.
**Is your feature request related to a problem? Please describe.**

The original DownloaderForReddit allowed the user to download videos/images from reddit posters who've linked imgur links in the comments. 

**Describe the solution you'd like**

Allow the option to download the media from commented imgur links, (original poster only).

**Additional context**

### LINKED ISSUE #76: Importing save files and user lists v2.3.2
On Win7.1, after every update to DFR, I'm greeted with a blank interface and have to manually re-add all my user lists. I'm sure I'm just misisng a step during the upgrade process--probably not a bug--but can anyone help me out with this? It's a pain to re-add hundreds of users over tens of lists. 

Right now I'm updating from v2.2.2 to v2.3.2. 

When I try to "Import Save File" and select \AppData\Roaming\SomeGuySoftware\DownloaderForReddit\save_file.dat, the program asks for permission to overwrite and then crashes (exits). At next start, it gives me the attached message box. "Save Permission Denied: 
A save file could not be created at this location...This likely means the application does not have permission..."
error

Thanks!

### LINKED ISSUE #98: Folders for Users vs Subreddits
How do I set the the main folder for all users to be saved in, and one for subreddits?

Right now both download in one folder.

### LINKED ISSUE #99: Import list
Would be nice to import a user or subreddit list through a textfile

### LINKED ISSUE #127: Add a User Name in the File Name
Dear developer. I would like to ask if it's possible to add an option to add the Reddit poster's user name into the filename right after post title. Something like this: POSTNAME_USERNAME.jpeg or mp4

Thank you

### PR DESCRIPTION
**Pull request by MalloyDelacroix**
This is the large version 3.0 update that has been underway for some time.  This update introduces the sqlite backend and robust ways to view the data stored in the database.  This update also brings large improvements in reliability, download speed, and user customization.

Since this is such a large change to the code base, this will be a beta version until it has been sufficiently tested by users.

Fixes #18
Fixes #22
Fixes #30
Fixes #46
Fixes #47
Fixes #72
Fixes #76
Fixes #98
Fixes #99
Fixes #127

### CODE REVIEW

**Comment by MalloyDelacroix** [comment]
I will leave this pull request open for a few days before merging for review of any critical issues that may have been overlooked.

**Comment by zacker150** [comment]
Link to build: https://github.com/MalloyDelacroix/DownloaderForReddit/actions/runs/181646028

File: DownloaderForReddit/gui/add_reddit_object_dialog.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        self.refresh_name_count()
+
+    def remove_object_from_list(self):
+        for index in self.multi_object_list_widget.selectedIndexes():
+            self.multi_object_list_widget.takeItem(index.row())
+        self.refresh_name_count()
+
+    def import_list(self):
+        file_path = self.get_import_file_path()
+        if file_path is not None:
```
Comment (phrasek) [inline DownloaderForReddit/gui/add_reddit_object_dialog.py:66]: This crashes if the input file is empty, adding a check for empty files fixes this.
Something like 
`if os.stat(file_path).st_size != 0`

Comment (MalloyDelacroix) [reply DownloaderForReddit/gui/add_reddit_object_dialog.py]: Nice find, thanks!

**Pull request merged**